### PR TITLE
remove special handling of custom event type

### LINKED ISSUE #61: Declare custom events without the ruma_event! macro
Custom events currently require some special-casing to work with `ruma_event!`. We sould get rid of this special-casing and instead declare them manually.

### PR DESCRIPTION
**Pull request by DevinR528**
Resolves #61 
Partially relates to #62

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -153,6 +151,12 @@ pub mod collections {
         pub mod only;
     }
 }
+///
+pub mod custom {
+    pub mod basic;
+    pub mod room;
+    pub mod state;
+}
```
Comment (DevinR528) [inline src/lib.rs]: Is the mod structure okay here?

File: src/lib.rs [hunk 2]
```diff
@@ -406,53 +411,53 @@ pub trait StateEvent: RoomEvent {
...
+//     use ruma_events_macros::ruma_event;
+//     use serde_json::Value;
+
+//     ruma_event! {
+//         /// A custom state event not covered by the Matrix specification.
+//         CustomStateEvent {
+//             kind: StateEvent,
+//             event_type: Custom,
+//             content_type_alias: {
+//                 /// The payload for `CustomStateEvent`.
```
Comment (DevinR528) [inline src/lib.rs]: TODO remove commented out code

File: ruma-events-macros/src/gen.rs [hunk 3]
```diff
@@ -181,9 +166,7 @@ impl ToTokens for RumaEvent {
             serialize_field_calls.push(serialize_field_call);
         }
 
-        let (manually_serialize_type_field, import_event_in_serialize_impl) = if self.is_custom {
-            (TokenStream::new(), TokenStream::new())
-        } else {
+        let (manually_serialize_type_field, import_event_in_serialize_impl) = {
```
Comment (jplatte) [inline ruma-events-macros/src/gen.rs]: There's no longer a reason to initialize these two together from a tuple.

File: src/custom/basic.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    type Raw = raw::CustomEvent;
+    fn from_raw(raw: raw::CustomEvent) -> Self {
+        Self {
+            content: FromRaw::from_raw(raw.content),
+            event_type: raw.event_type,
+        }
+    }
+}
+
+impl Serialize for CustomEvent {
```
Comment (jplatte) [inline src/custom/basic.rs]: This impl can be derived by serde if the `event_type` field has `#[rename("type")]`.

File: src/custom/basic.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    /// The event's content.
+    pub content: CustomEventContent,
+    /// The custom type of the event.
+    pub event_type: String,
+}
+
+/// The payload for `CustomEvent`.
+pub type CustomEventContent = Value;
+
+impl FromRaw for CustomEvent {
```
Comment (jplatte) [inline src/custom/basic.rs]: This impl can be derived, that's part of the reason I split `FromRaw` deriving from `ruma_event!`.

File: src/custom/basic.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    /// The event's content.
+    fn content(&self) -> &Self::Content {
+        &self.content
+    }
+    /// The type of the event.
+    fn event_type(&self) -> EventType {
+        EventType::Custom(self.event_type.clone())
+    }
+}
+/// "Raw" versions of the event and its content which implement `serde::Deserialize`.
```
Comment (jplatte) [inline src/custom/basic.rs]:
```suggestion
}

/// "Raw" versions of the event and its content which implement `serde::Deserialize`.
```

File: src/custom/basic.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+pub(crate) mod raw {
+    use super::*;
+    /// A custom basic event not covered by the Matrix specification.
```
Comment (jplatte) [inline src/custom/basic.rs]:
```suggestion
    use super::*;

    /// A custom event not covered by the Matrix specification.
```

File: src/custom/room.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
+//! A custom room event outside of the Matrix specification.
+
+use crate::{Event, EventType, FromRaw, RoomEvent};
+use serde::{Deserialize, Serialize, Serializer};
+use serde_json::Value;
+/// A custom room event not covered by the Matrix specification.
```
Comment (jplatte) [inline src/custom/room.rs]:
```suggestion
use serde_json::Value;

/// A custom room event not covered by the Matrix specification.
```

File: src/lib.rs (hunk 1 above)
Comment (jplatte) [reply src/lib.rs]: Why make a breaking change instead of keeping the current modules?

File: tests/ruma_events_macros.rs [hunk 9]
```diff
@@ -121,18 +121,7 @@ mod common_case {
 
 mod custom_event_type {
```
Comment (jplatte) [inline tests/ruma_events_macros.rs]: This test module can be removed completely as it was testing the codegen for custom events.

File: src/lib.rs (hunk 1 above)
Comment (DevinR528) [reply src/lib.rs]: I thought there would be more code in each section but with all the derives its pretty short. will fix

File: src/lib.rs [hunk 10]
```diff
@@ -406,53 +406,265 @@ pub trait StateEvent: RoomEvent {
...
+        /// A key that determines which piece of room state the event represents.
+        pub state_key: String,
+        /// Additional key-value pairs not signed by the homeserver.
+        pub unsigned: Option<serde_json::Value>,
+    }
+
+    /// The payload for `CustomStateEvent`.
+    pub type CustomStateEventContent = Value;
+
+    impl StateEvent for CustomStateEvent {
```
Comment (jplatte) [inline src/lib.rs]: Could you move this impl after the `RoomEvent` one?

File: src/lib.rs [hunk 11]
```diff
@@ -406,53 +406,265 @@ pub trait StateEvent: RoomEvent {
...
+        /// The unique identifier for the user who sent this event.
+        pub sender: ruma_identifiers::UserId,
+        /// Additional key-value pairs not signed by the homeserver.
+        pub unsigned: Option<serde_json::Value>,
+    }
+
+    /// The payload for `CustomRoomEvent`.
+    pub type CustomRoomEventContent = Value;
+
+    impl RoomEvent for CustomRoomEvent {
```
Comment (jplatte) [inline src/lib.rs]: Could you move this impl after the `Event` one?

**Comment by jplatte** [comment]
Thanks! Since I thought the patch I sent to you was part of this PR I added `Co-authored-by` in the squash commit (but the patch is part of #64, not this PR). I have reverted that commit and pushed another one without `Co-authored-by`.

**Comment by DevinR528** [comment]
Thanks, in the future you don't have to worry about it if it's more work.

**Comment by jplatte** [comment]
It would have been trivial, would I be able to disable branch protections. But it still wasn't much work, and I was just fixing what I broke, plus I consider authorship information important.

**Pull request merged**